storage/inmem/txn.go

```diff
@@ -0,0 +1,410 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package inmem
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/open-policy-agent/opa/storage"
+)
+
+type transaction struct {
+	xid      uint64
+	write    bool
+	log      *snapshot
+	db       *store
+	policies map[string][]byte
+}
+
+type snapshot struct {
+	path   storage.Path
+	remove bool
+	stale  bool
+	value  interface{}
+	next   *snapshot
+	prev   *snapshot
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) assert(op storage.PatchOp, path storage.Path, value interface{}) error {
+
+	if !txn.write {
+		return &storage.Error{
+			Code:    storage.InvalidTransactionErr,
+			Message: "data write during read transaction",
+		}
+	}
+
+	if len(path) == 0 {
+		return txn.assertRoot(op, value)
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if curr.path.Equal(path) {
+			if curr.remove {
+				if op != storage.AddOp {
+					return notFoundError(path)
+				}
+			}
+			curr.stale = true
+			break
+		} else if curr.path.HasPrefix(path) {
+			curr.stale = true
+		} else if path.HasPrefix(curr.path) {
+			if curr.remove {
+				return notFoundError(path)
+			}
+			suffix := path[len(curr.path):]
+			path, remove, value, err := validate(curr.value, op, suffix, 0, value)
+			if err != nil {
+				return err
+			}
+			if len(path) == 0 {
+				curr.remove = remove
+				curr.value = value
+			} else {
+				apply(curr.value, path, remove, value)
+			}
+			return nil
+		}
+	}
+
+	path, remove, value, err := validate(txn.db.data, op, path, 0, value)
+	if err != nil {
+		return err
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if curr.stale {
+			if curr == txn.log {
+				txn.log = curr.next
+			}
+			if curr.next != nil {
+				curr.next.prev = curr.prev
+			}
+			if curr.prev != nil {
+				curr.prev.next = curr.next
+			}
+		}
+	}
+
+	curr := &snapshot{
+		path:   path,
+		remove: remove,
+		value:  value,
+		next:   txn.log,
+		prev:   nil,
+	}
+
+	if txn.log != nil {
+		txn.log.prev = curr
+	}
+
+	txn.log = curr
+	return nil
+}
+
+func (txn *transaction) assertRoot(op storage.PatchOp, value interface{}) error {
+	if op == storage.RemoveOp {
+		return invalidPatchError(rootCannotBeRemovedMsg)
+	}
+	if _, ok := value.(map[string]interface{}); !ok {
+		return invalidPatchError(rootMustBeObjectMsg)
+	}
+	txn.log = &snapshot{
+		path:   storage.Path{},
+		remove: false,
+		value:  value,
+		next:   nil,
+	}
+	return nil
+}
+
+func (txn *transaction) commit() (result storage.TriggerEvent) {
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if len(curr.path) > 0 {
+			apply(txn.db.data, curr.path, curr.remove, curr.value)
+		} else {
+			txn.db.data = curr.value.(map[string]interface{})
+		}
+		result.SetDataChanged()
+	}
+	for id, bs := range txn.policies {
+		if bs == nil {
```

Comment:
this use of nil as a sign of remove should be mentioned in snapshot struct comment.